src/aleph/http/client.clj

```diff
@@ -386,95 +392,314 @@
+                  (s/close! body-stream))
+                (handle-error (Http2Exception. (Http2Error/valueOf error-code) msg))))))]
+
+    (netty/channel-inbound-handler
+
+      :exception-caught
+      ([_ ctx ex]
+       (handle-error ex))
+
+      :channel-inactive
+      ([_ ctx]
+       (s/close! body-stream)
+       (.fireChannelInactive ctx))
+
+      :channel-read
+      ([_ ctx msg]
+       (cond
+         (instance? Http2HeadersFrame msg)
+         (let [headers (.headers ^Http2HeadersFrame msg)
+               ring-resp {:status            (->> headers (.status) (.convertToInt netty/char-seq-val-converter)) ; might be an AsciiString
+                          :headers           (http2/netty-http2-headers->map headers)
+                          :aleph/keep-alive? true           ; not applicable to HTTP/2, but here for compatibility
+                          :aleph/complete    complete
+                          :body              body-stream}]
+           (d/success! response-d ring-resp)
+           (when (.isEndStream ^Http2HeadersFrame msg)
+             (s/close! body-stream)))
+
+         (instance? Http2DataFrame msg)
+         (let [content (.content ^Http2DataFrame msg)]
+           (if raw-stream?
+             (netty/put! (.channel ctx) body-stream content)
+             (do
+               (netty/put! (.channel ctx) body-stream (netty/buf->array content))
+               (.release ^ReferenceCounted msg)))
+           (when (.isEndStream ^Http2DataFrame msg)
+             (s/close! body-stream)))
+
+         :else
+         (.fireChannelRead ctx msg)))
+
+      :user-event-triggered
+      ([_ ctx evt]
+       (handle-shutdown-frame evt)
+       (.fireUserEventTriggered ctx evt)))))
+
+
+(defn- build-http2-stream-pipeline
+  "Set up the pipeline for an HTTP/2 stream channel"
+  [^Http2StreamChannel ch resp proxy-options logger pipeline-transform raw-stream? response-buffer-size]
+  (log/trace "build-http2-stream-pipeline called")
+
+  (let [p (.pipeline ch)]
+    ;; necessary for multipart support in HTTP/2
+    #_(.addLast p
+                "stream-frame-to-http-object"
+                ^ChannelHandler stream-frame->http-object-codec)
+    (.addLast p
+              "streamer"
+              ^ChannelHandler (ChunkedWriteHandler.))
+    (.addLast p
+              "handler"
               ^ChannelHandler
-              (pending-proxy-connection-handler response-stream)))))
-      (when (some? logger)
-        (.addFirst pipeline "activity-logger" ^ChannelHandler logger))
-      (pipeline-transform pipeline))))
+              (http2-client-handler ch resp raw-stream? response-buffer-size))
+
+    (when (some? proxy-options)
+      (throw (IllegalArgumentException. "Proxying HTTP/2 messages not supported yet")))
+
+    (add-non-http-handlers p logger pipeline-transform)
+
+    (log/trace "added all stream-chan handlers")
+    (log/debug "Stream chan:" ch)
+
+    ch))
+
+(defn- h2-stream-chan-initializer
```

Comment:
This is not used.

Reply:
True...it was being used for `server-initiated-stream-chan-initializer`, which I replaced with a placeholder that just logs an error if the server initiates a stream. I was using for the client, but stopped because I needed different code to snc the manifold and netty side of things.

I thought it might still be useful for the server-side, so I haven't deleted it yet.